doc/api/module.md

```diff
@@ -66,6 +66,47 @@ const require = createRequire(import.meta.url);
 const siblingModule = require('./sibling-module');
 ```
 
+### `module.detectSyntax(code)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.0 - Early development
+
+The `module.detectSyntax()` method attempts to determine the syntax of a given source code string.
+This detection is performed only by parsing the source code; there is no evaluation or reading of file extensions
+or reading of `package.json` files.
+Only TypeScript syntax that can be erased is supported.
+
+* `code` {string} The code to detect the syntax of.
+* Returns: {string|undefined} The detected syntax of the code. Possible values are:
+  * `'commonjs'` for CommonJS module code.
+  * `'module'` for ECMAScript module code.
+  * `'commonjs-typescript'` for CommonJS TypeScript code.
+  * `'module-typescript'` for ECMAScript TypeScript code.
+  * `undefined` if the syntax cannot be detected.
+    **Default:** `'commonjs'`.
+
+```mjs
+import { detectSyntax } from 'node:module';
+detectSyntax('export const a = 1;'); // 'module'
+detectSyntax('const a = 1;'); // 'commonjs'
+detectSyntax('export const a: number = 1;'); // 'module-typescript'
+detectSyntax('const a: number = 1;'); // 'commonjs-typescript'
+detectSyntax('const foo;'); // Invalid syntax, returns undefined
+```
```

Comment:
Maybe this should instead return two booleans where one indicates if the code contained module specific syntax and the other whether it contained typescript syntax? Stating `const a = 1;` is commonjs isn't exactly accurate because that can't be determined solely from the text, as alluded to in the description.